File: Rating.cs
using System;

namespace BookRecommender.Models
{
    /// <summary>
    /// Класс оценки товара
    /// </summary>
    public class Rating
    {
        private static int _nextId = 1;

        public int Id { get; private set; }
        public User User { get; private set; }
        public Product Product { get; private set; }
        public int Score { get; private set; }
        public DateTime Timestamp { get; private set; }

        public Rating() { }

        public Rating(int id, User user, Product product, int score, DateTime? timestamp = null)
        {
            if (user == null)
                throw new ArgumentNullException(nameof(user));
            if (product == null)
                throw new ArgumentNullException(nameof(product));
            if (score < 1 || score > 5)
                throw new ArgumentOutOfRangeException(nameof(score), "Score must be 1-5");

            Id = id;
            User = user;
            Product = product;
            Score = score;
            Timestamp = timestamp ?? DateTime.Now;
        }

        public Rating(User user, Product product, int score, DateTime? timestamp = null)
        {
            if (user == null)
                throw new ArgumentNullException(nameof(user));
            if (product == null)
                throw new ArgumentNullException(nameof(product));
            if (score < 1 || score > 5)
                throw new ArgumentOutOfRangeException(nameof(score), "Score must be 1-5");

            Id = _nextId++;
            User = user;
            Product = product;
            Score = score;
            Timestamp = timestamp ?? DateTime.Now;
        }

        public static void ResetIdCounter(int start = 1) => _nextId = start;

        public override string ToString() =>
            $"Rating #{Id}: {User.Name} → {Product.Name} = {Score}";
    }
}

File: Rating.cpp
#include "Rating.h"
#include "User.h"
#include "Product.h"
#include <stdexcept>

Rating::Rating() : id_(0), score_(0), timestamp_(0) {}

Rating::Rating(int id, std::shared_ptr<User> user, std::shared_ptr<Product> product,
    int score, std::time_t timestamp)
    : id_(id), user_(std::move(user)), product_(std::move(product)), score_(score), timestamp_(timestamp) {
    if (!user || !product) {
        throw std::invalid_argument("user and product must not be null");
    }
    if (score < 1 || score > 5) {
        throw std::invalid_argument("score must be 1–5");
    }
}

int Rating::getId() const { return id_; }
std::shared_ptr<User> Rating::getUser() const { return user_; }
std::shared_ptr<Product> Rating::getProduct() const { return product_; }
int Rating::getScore() const { return score_; }
std::time_t Rating::getTimestamp() const { return timestamp_; }
Rating::operator int() const { return score_; }